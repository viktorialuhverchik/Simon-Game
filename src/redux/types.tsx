export const POWER_OFF = "POWER_OFF";
export const POWER_ON = "POWER_ON";
export const TOGGLE_MODE = "TOGGLE_MODE";
export const START_GAME = "START_GAME";
export const MOVE_GAME = "MOVE_GAME";
export const MOVE_USER = "MOVE_USER";
export const UPDATE_SCORE = "UPDATE_SCORE";
export const TOGGLE_AVAILABLE = "TOGGLE_AVAILABLE";