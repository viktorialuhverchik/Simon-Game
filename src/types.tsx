export interface State {
    app: {
        isOn: boolean
        isStart: boolean
        isStrictMode: boolean
    },
    game: {
        movesGame: number[]
        score: number
    }
};

export interface PropsDisplay {
    isOn: boolean
    isStart: boolean
    score: number
};

export interface PropsGame {
    isStart: boolean
    isStrictMode: boolean
    movesGame: number[]
    score: number
};

export interface PropsSettings {
    isOn: boolean
    isStrictMode: boolean
    isStart: boolean
    score: number
};

export interface IThemes {
    classical: string
    synthwave: string
};
