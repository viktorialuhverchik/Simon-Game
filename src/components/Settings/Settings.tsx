import React, { FC } from 'react';
import { useDispatch } from 'react-redux';
import Display from '../Display/Display';
import {
    powerOff,
    powerOn, 
    toggleStart,
    toggleMode
} from '../../redux/actions/actions';
import { PropsSettings } from '../../types';

import './Settings.css';

const Settings: FC<PropsSettings> = ({ isOn, isStrictMode, isStart, score }) => {

    const dispatch = useDispatch();

    return (
        <div className="setting-panel">
            <h1>Simon</h1>
            <div className="button-wrapper">

                <div className="power-button_wrapper">
                    <div
                        className={`power-button available ${isOn ? "pushed-button" : ""}`}
                        onClick={() => !isOn ? dispatch(powerOn()) : dispatch(powerOff(isStart, isStrictMode, dispatch))}
                        data-testid="power"
                    >
                    </div>
                    {!isOn ? <span>ON</span> : <span>OFF</span>}
                </div>

                <div className="modes-button_wrapper">
                    <div
                        className={`modes-button ${!isOn || isStart ? "disabled" : "available"} ${isStrictMode ? "pushed-button" : ""}`}
                        onClick={() => dispatch(toggleMode(isStrictMode))}
                        data-testid="mode"
                    >
                    </div>
                    {!isStrictMode ? <span>SIMPLE</span> : <span>STRICT</span>}
                </div>

                <div className="start-button_wrapper">
                    <div
                        className={`toggle-start_button ${!isOn ? "disabled" : "available"} ${isStart ? "pushed-button" : ""}`}
                        onClick={() => dispatch(toggleStart(!isStart, 0, []))}
                        data-testid="start"
                    >
                    </div>
                    {!isStart ? <span>START</span> : <span>STOP</span>}
                </div>

            </div>
            <Display isOn={isOn} isStart={isStart} score={score}/>
        </div>
    );
};

export default Settings;
