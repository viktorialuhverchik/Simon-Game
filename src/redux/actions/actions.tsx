import React from 'react';
import {
    POWER_ON,
    POWER_OFF,
    TOGGLE_MODE,
    START_GAME,
    MOVE_GAME,
    UPDATE_SCORE, 
    TOGGLE_AVAILABLE,
    MOVE_USER
} from "../types";

export const powerOn = () => ({
    type: POWER_ON
});

export const powerOff = (isStart: boolean, score: number, dispatch: any, isAvailableClick: boolean) => {
    if(isStart) {
        dispatch(startGame(isStart));
    };
    score = 0;
    dispatch(updateScore(score));
    dispatch ({
        type: MOVE_GAME,
        moveGame: []
    });
    return {
        type: POWER_OFF
    };
};

export const sayHello = (isOn: boolean, isStart: boolean) => (
    !isOn || isStart ? "" : <h6>Hello! <br/> Select mode and press start</h6>
);

export const toggleMode = (isStrictMode: boolean) => ({
    type: TOGGLE_MODE,
    isStrictMode: !isStrictMode
});

export const startGame = (isStart: boolean) => ({
    type: START_GAME,
    isStart: !isStart
});

export const addMoveGame = (move: any) => {
    let nextStep: number = Math.round(Math.random() * 3);
    move.push(nextStep);
    console.log(move);
    return {
        type: MOVE_GAME,
        moveGame: move
    };
};

// export const addMoveUser = (step: number) => {
//     let moveUser: any = [];
//     moveUser.push(step);
//     console.log(moveUser);
//     return {
//         type: MOVE_USER,
//         moveUser
//     };
// };

export const setAvailableClick = (isAvailableClick: boolean) => ({
    type: TOGGLE_AVAILABLE,
    isAvailableClick: !isAvailableClick
});

export const updateScore = (score: number) => ({
    type: UPDATE_SCORE,
    score
});

export const compareMove = (isStrictMode: boolean, moveGame: any, moveUser: any, score: number) => {
    return (dispatch: any) => {
        // if(moveGame.length !== moveUser.length) {
        //     console.log("Loose");
        //     if(isStrictMode) {
        //         score = 0;
        //         dispatch(updateScore(score));
        //     };
        //     return;
        // };
        for (let i = 0; i < moveGame.length; i++) {
            if(moveGame[i] !== moveUser) {
                console.log("Loose");
                if(isStrictMode) {
                    score = 0;
                    dispatch(updateScore(score));
                };
            } else {
                console.log("Win");
                score++;
                console.log(score);
                dispatch(updateScore(score));
            };
        };
    };
};
