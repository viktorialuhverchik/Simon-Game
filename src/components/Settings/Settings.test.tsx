import React from 'react';
import { renderWithRedux } from '../../index.test';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { fireEvent } from '@testing-library/react';
import { powerOff, powerOn, toggleMode, toggleStart } from '../../redux/actions/actions';
import { POWER_OFF, POWER_ON, TOGGLE_MODE, TOGGLE_START, UPDATE_SCORE } from '../../redux/types';
import Settings from './Settings';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Settings component',() => {

    let store: any;

    beforeEach(() => {
        store = mockStore();
        fetchMock.resetMocks();
    });

    afterEach(() => {
        fetchMock.mockClear();
    });

    it('renders Settings component', () => {
        const wrapper = renderWithRedux(<Settings />);
        expect(wrapper).toMatchSnapshot();
    });

    it('should power on', () => {
        const expectedActions = { type: POWER_ON };
        store.dispatch(powerOn());
        expect(store.getActions()[0]).toEqual(expectedActions);
    });

    it('should power off', () => {
        let isStart: boolean = true;
        let isStrictMode: boolean = false; 
        let dispatch: any = store.dispatch;
        const expectedActions = { type: POWER_OFF };
        const expectedActions1 = { type: UPDATE_SCORE, score: 0 }; 
        store.dispatch(powerOff(isStart, isStrictMode, dispatch));
        expect(store.getActions()[0]).toEqual(expectedActions && expectedActions1);
    });

    it('should toggle mode', () => {
        let isStrictMode: boolean = false;
        const expectedActions = { type: TOGGLE_MODE, isStrictMode: true };
        store.dispatch(toggleMode(isStrictMode));
        expect(store.getActions()[0]).toEqual(expectedActions);
    });

    it('should toggle start', () => {
        let isStart: boolean = false;
        let score: number = 0; 
        let movesGame: any = [];
        const expectedActions = { type: TOGGLE_START, isStart: true };
        const expectedActions1 = { type: UPDATE_SCORE, score: 0 };   
        store.dispatch(toggleStart(isStart, score, movesGame));
        expect(store.getActions()[0]).toEqual(expectedActions && expectedActions1);
    });

    it('click on power', () => {
        const { getByTestId } = renderWithRedux(<Settings/>);
        fireEvent.click(getByTestId("power"));
        expect(getByTestId("power")).toBeTruthy();
    });

    it('click on mode', () => {
        const { getByTestId } = renderWithRedux(<Settings/>);
        fireEvent.click(getByTestId("mode"));
        expect(getByTestId("mode")).toBeTruthy();
    });

    it('click on start', () => {
        const { getByTestId } = renderWithRedux(<Settings/>);
        fireEvent.click(getByTestId("start"));
        expect(getByTestId("start")).toBeTruthy();
    });
});