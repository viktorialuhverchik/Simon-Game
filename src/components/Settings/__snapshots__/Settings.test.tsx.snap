// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings component renders Settings component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="setting-panel"
      >
        <h1>
          Simon
        </h1>
        <div
          class="button-wrapper"
        >
          <div
            class="power-button_wrapper"
          >
            <div
              class="power-button available "
              data-testid="power"
            />
            <span>
              ON
            </span>
          </div>
          <div
            class="modes-button_wrapper"
          >
            <div
              class="modes-button disabled "
              data-testid="mode"
            />
            <span>
              SIMPLE
            </span>
          </div>
          <div
            class="start-button_wrapper"
          >
            <div
              class="toggle-start_button disabled "
              data-testid="start"
            />
            <span>
              START
            </span>
          </div>
        </div>
        <div
          class="app-display"
        >
          
          
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="setting-panel"
    >
      <h1>
        Simon
      </h1>
      <div
        class="button-wrapper"
      >
        <div
          class="power-button_wrapper"
        >
          <div
            class="power-button available "
            data-testid="power"
          />
          <span>
            ON
          </span>
        </div>
        <div
          class="modes-button_wrapper"
        >
          <div
            class="modes-button disabled "
            data-testid="mode"
          />
          <span>
            SIMPLE
          </span>
        </div>
        <div
          class="start-button_wrapper"
        >
          <div
            class="toggle-start_button disabled "
            data-testid="start"
          />
          <span>
            START
          </span>
        </div>
      </div>
      <div
        class="app-display"
      >
        
        
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
